src/vmware/vmxnet3_net.c

```diff
@@ -0,0 +1,628 @@
+    assert(vxrxq_rs->comp_ring_len == VMXNET3_MAX_RX_NCOMPDESC);
+    assert(vxrxq_rs->comp_ring_len == VMXNET3_MAX_RX_NCOMPDESC);
+    assert(vxrxq_rs->driver_data_len == sizeof(*dev->vmx_rxq[0]));
+    assert(vxrxq_rs->intr_idx == 1);
+    assert(dev->vmx_rxq[0]->vxrxq_intr_idx == 1);
+
+    assert(vxrxq_rs->stopped == 0);
+    assert(vxrxq_rs->error == 0);
+
+    //rx bufs
+    for(int i = 0; i < VMXNET3_RXRINGS_PERQ; ++i) {
+        struct vmxnet3_rxring *rxr = &dev->vmx_rxq[0]->vxrxq_cmd_ring[i];
+        assert(rxr->vxrxr_ndesc == VMXNET3_MAX_RX_NDESC);
+        assert(rxr->vxrxr_gen == (VMXNET3_INIT_GEN ^ 1));
+        assert(rxr->vxrxr_paddr == physical_from_virtual(rxr->vxrxr_rxd));
+        assert(rxr->vxrxr_desc_skips == 0);
+        assert(rxr->vxrxr_refill_start == 0);
+        for(int j = 0; j<rxr->vxrxr_ndesc; ++j) {
+            assert(rxr->vxrxr_rxd[j].addr == physical_from_virtual(dev->rx_pbuf[i][j]->payload));
+            assert(rxr->vxrxr_rxd[j].btype == (i == 0 ? VMXNET3_BTYPE_HEAD : VMXNET3_BTYPE_BODY));
+            assert(rxr->vxrxr_rxd[j].dtype == 0);
+            assert(rxr->vxrxr_rxd[j].len == vn->rxbuflen);
+            assert(rxr->vxrxr_rxd[j].gen == VMXNET3_INIT_GEN);
+        }
+    }
+
+    //rx descs buf
+    struct vmxnet3_comp_ring *rxc = &dev->vmx_rxq[0]->vxrxq_comp_ring;
+    assert(rxc->vxcr_next == 0);
+    assert(rxc->vxcr_gen == VMXNET3_INIT_GEN);
+    assert(rxc->vxcr_ndesc == VMXNET3_MAX_RX_NCOMPDESC);
+    for(int j = 0; j < rxc->vxcr_ndesc; ++j) {
+        struct vmxnet3_rxcompdesc* rxcd = rxc->vxcr_u.rxcd;
+        assert(rxcd->rxd_idx == 0);
+        assert(rxcd->len == 0);
+    }
+#endif
+}
+
+static void
+vmxnet3_net_attach(heap general, heap page_allocator, pci_dev d)
+{
+    struct vmxnet3_pci *dev = allocate(general, sizeof(struct vmxnet3_pci));
+
+    dev->_dev = *d;
+    dev->dev = &dev->_dev;
+
+    pci_bar_init(dev->dev, &dev->bar0, 0, 0, -1);
+    pci_bar_init(dev->dev, &dev->bar1, 1, 0, -1);
+
+    pci_set_bus_master(dev->dev);
+    pci_enable_msix(dev->dev);
+
+    dev->general = general;
+    dev->contiguous = page_allocator;
+
+    vmxnet3_write_cmd(dev, VMXNET3_CMD_DISABLE);
+    vmxnet3_write_cmd(dev, VMXNET3_CMD_RESET);
+
+    vmxnet3 vn = allocate(dev->general, sizeof(struct vmxnet3));
+    vn->dev = dev;
+    vn->n = allocate(dev->general, sizeof(struct netif));
+
+    vn->rxbuflen = VMXNET3_RX_MAXSEGSIZE;
+    vn->rxbuffers = allocate_objcache(dev->general, page_allocator,
+                      vn->rxbuflen + sizeof(struct xpbuf), PAGESIZE_2M);
+    spin_lock_init(&vn->rx_buflock);
+
+    dev->vmx_ds = allocate_zero(dev->contiguous, sizeof(struct vmxnet3_driver_shared));
+
+    vn->rx_servicequeue = allocate_queue(dev->general, VMXNET3_RX_SERVICEQUEUE_DEPTH);
+    assert(vn->rx_servicequeue != INVALID_ADDRESS);
+    vn->rx_service = closure(dev->general, vmxnet3_rx_service_bh, vn);
+
+    vn->rx_intr_handler = closure(dev->general, rx_interrupt, vn);
```

Comment:
I don't see where vn->tx_intr_handler gets initialized. Am I overlooking something, or is the tx int handler missing?

Reply:
It is not used (it is disabled). I will either remove pci_setup_msix() for tx_intr_handler or h`ve an empty handler.